lib/foreman_inventory_upload/generators/fact_helpers.rb

```diff
@@ -57,55 +57,187 @@ def cloud_provider(host)
+          begin
+            parsed_insights_array = JSON.parse(fact_value(host, 'insights_client::obfuscated_hostname') || '[]')
+          rescue JSON::ParserError
+            # If the fact contains invalid JSON, default to an empty array to prevent upload failures.
+            parsed_insights_array = []
+          end
+
+          # Find the specific item in the parsed array where the 'original' hostname matches
+          # the host's actual FQDN.
+          parsed_insights_item = parsed_insights_array.find { |item| item['original'] == host.fqdn }
+
+          # Return the 'obfuscated' value from the found item if it exists.
+          # The `&.` (safe navigation operator) prevents an error if `parsed_insights_item` is nil.
+          # If no matching obfuscated hostname is found in the fact, or if the fact itself was empty,
+          # then dynamically obfuscate the host's FQDN using the `obfuscate_fqdn` helper method.
+          parsed_insights_item&.[]('obfuscated') || obfuscate_fqdn(host.fqdn)
+        else
+          # If hostname obfuscation is not enabled for this host, return the host's original FQDN.
+          host.fqdn
+        end
       end
 
       def obfuscate_fqdn(fqdn)
         "#{Digest::SHA1.hexdigest(fqdn)}.example.com"
       end
 
       def obfuscate_ips?(host)
-        insights_client_setting = fact_value(host, 'insights_client::obfuscate_ip_enabled')
-        insights_client_setting = ActiveModel::Type::Boolean.new.cast(insights_client_setting)
-        return insights_client_setting unless insights_client_setting.nil?
-
-        Setting[:obfuscate_inventory_ips]
+        # Returns true if IP obfuscation should be applied for a given host.
+        # This method determines whether IP addresses associated with a host
+        # should be hidden or masked based on global settings and host-specific facts.
+
+        # Check the global setting for IP obfuscation.
+        # If 'obfuscate_inventory_ips' is set to true in the application settings,
+        # then IP obfuscation is always enabled, regardless of host-specific settings.
+        return true if Setting[:obfuscate_inventory_ips]
+
+        # Retrieve host-specific facts related to IP obfuscation from 'insights_client'.
+        # These facts indicate whether IPv4 and IPv6 obfuscation are enabled for this particular host.
+        insights_client_ipv4_setting = fact_value(host, 'insights_client::obfuscate_ipv4_enabled')
+        insights_client_ipv6_setting = fact_value(host, 'insights_client::obfuscate_ipv6_enabled')
+
+        # Cast the retrieved settings to boolean values.
+        # This ensures that the values from the facts (which might be strings or other types)
+        # are correctly interpreted as true or false.
+        casted_ipv4_setting = ActiveModel::Type::Boolean.new.cast(insights_client_ipv4_setting)
+        casted_ipv6_setting = ActiveModel::Type::Boolean.new.cast(insights_client_ipv6_setting)
```

Comment:
I don't think "casted" is a word

```suggestion
        cast_ipv4_setting = ActiveModel::Type::Boolean.new.cast(insights_client_ipv4_setting)
        cast_ipv6_setting = ActiveModel::Type::Boolean.new.cast(insights_client_ipv6_setting)
```